sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -515,3 +515,96 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
 
   override def prettyName: String = "str_to_map"
 }
+
+/**
+ * Adds/replaces field in struct by name.
+ *
+ * @param children Seq(struct, name, val)
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(struct, name, val) - Adds/replaces field in struct by name.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(NAMED_STRUCT("a", 1, "b", 2), "c", 3);
+       {"a":1,"b":2,"c":3}
+      > SELECT _FUNC_(NAMED_STRUCT("a", 1, "b", 2), "b", 3);
+       {"a":1,"b":3}
+      > SELECT _FUNC_(CAST(NULL AS struct<a:int,b:int>), "c", 3);
+       {"a":null,"b":null,"c":3}
```

Comment:
this is a bit counter-intuitive. Most SQL functions return null if the input is null. Do we have a valid real-world use case to justify this behavior?

Reply:
> Most SQL functions return null if the input is null. 

Agreed that this is true in the majority of cases.

> Do we have a valid real-world use case to justify this behavior?

The justification for this behaviour is not based on any use case. 
Originally, the code in this PR did exactly what you're saying i.e. null input returns null output. 
However, based on some feedback on the PR, I switched to the current behaviour for both simplicity of implementation and easy optimization of result plans. Please refer to the discussion [here](https://github.com/apache/spark/pull/27066#issuecomment-609512144) and [here](https://github.com/apache/spark/pull/27066#issuecomment-609671222) for more details. 
IMO this is an acceptable compromise. 
The user always has the option of the wrapping this method in a case-when expression themselves.

Reply:
does this apply to the inner struct as well?
```
with_field(
  NAMED_STRUCT("a", 1, "b", CAST(NULL AS struct<a:int,b:int>)),
  ''b.c",
  3)
```

Reply:
Yes indeed. 
```
-- !query
SELECT WITH_FIELD(NAMED_STRUCT("a", 1, "b", CAST(NULL AS struct<a:int,b:int>)), "b.c", 3)
-- !query output
{"a":1,"b":{"a":null,"b":null,"c":3}}
```